### PR TITLE
POC: integration test framework for GitHub Actions

### PR DESCRIPTION
**Pull request by sebrandon1**
## Summary
- Adds a shell-based integration test harness that exercises the compiled recert binary directly in GitHub Actions
- 20 test scenarios covering config validation, filesystem crypto operations across all supported key types (RSA 2048/4096, ECDSA P-256/P-384, cross-algorithm chains), cert-manager certificate handling, Ed25519 rejection, and etcd connectivity
- Verifies that recert preserves key algorithms, sizes, and curves after regeneration
- Includes a one-line fix in `src/recert.rs` to skip `is_encryption_enabled` when no etcd endpoint is configured, enabling filesystem-only operation

## Test Scenarios

| # | Scenario | What it tests |
|---|---|---|
| 1 | `test_config_cli_vs_env` | RECERT_CONFIG + CLI args rejected |
| 2 | `test_config_ip_change_only` | ip_change_only without cn_san_replace rejected |
| 3 | `test_config_minimal_valid` | Minimal config with dry_run succeeds |
| 4 | `test_config_mutual_exclusion` | 4 mutually exclusive flag combos rejected |
| 5 | `test_config_unknown_key` | Unknown YAML keys rejected |
| 6 | `test_crypto_all_key_types` | All 10 cert/key pairs (RSA+ECDSA+cross) regenerated, algorithms preserved |
| 7 | `test_crypto_cn_san_replace` | DNS SAN replacement works |
| 8 | `test_crypto_cross_algorithm` | EC leaf under RSA CA + RSA leaf under EC CA |
| 9 | `test_crypto_dry_run` | Files unchanged in dry-run mode |
| 10 | `test_crypto_ecdsa_p256` | P-256 CA + leaf regeneration, curve preserved |
| 11 | `test_crypto_ecdsa_p384` | P-384 CA + leaf regeneration, curve preserved |
| 12 | `test_crypto_ed25519_rejected` | Ed25519 certs properly rejected (positive test after PR 1817 merges) |
| 13 | `test_crypto_force_expire` | Cert notAfter set to past |
| 14 | `test_crypto_rsa2048` | RSA-2048 CA + leaf regeneration, key size preserved |
| 15 | `test_crypto_rsa4096` | RSA-4096 CA + leaf regeneration, key size preserved |
| 16 | `test_crypto_scan_dir` | Basic dir scan with regeneration verification |
| 17 | `test_crypto_summary_redaction` | Clean summary has no private key content |
| 18 | `test_etcd_bad_endpoint` | Bad etcd endpoint produces connection error |
| 19 | `test_etcd_cert_manager` | cert-manager Certificate CRD domain rename via etcd + stale CertificateRequest cleanup |
| 20 | `test_etcd_connect` | Full recert run against seeded etcd with OCP resources |

## Test plan
- [ ] All 20 integration test scenarios pass in CI
- [x] Existing rust.yml and yamllint.yml workflows unaffected
- [ ] Review recert code fix in src/recert.rs (one-line guard for no-etcd operation)

## Summary by CodeRabbit

- **New Features**
  - Added a containerized integration test suite covering configuration validation, certificate renewal, key and certificate replacement, expiration handling, SAN updates, summary redaction, and etcd workflows.
  - Added coverage for RSA and ECDSA certificates, certificate chains, dry-run behavior, post-processing-only mode, and unsupported key types.

- **Bug Fixes**
  - Improved encryption setup reliability when no etcd client is available.

- **Chores**
  - Added automated integration-test execution for pull requests and pushes, with failure artifacts retained for troubleshooting.

### CODE REVIEW

File: .github/workflows/integration.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+jobs:
+  integration:
+    name: Integration Tests
+    runs-on: ubuntu-latest
+    env:
+      ETCD_VERSION: v3.5.17
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: recursive
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply .github/workflows/integration.yml:22]: Fixed — added `permissions: contents: read` at workflow level and `persist-credentials: false` on the checkout step.

File: tests/integration/run_tests.sh [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+EXCLUDE_TAG=""
+while [[ $# -gt 0 ]]; do
+    case "$1" in
+        --tag)
+            TAG_FILTER="$2"
+            shift 2
+            ;;
+        --exclude-tag)
+            EXCLUDE_TAG="$2"
+            shift 2
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply tests/integration/run_tests.sh:31]: Valid edge case, but this script is only invoked by the workflow with controlled arguments. The `set -u` error message is clear enough for debugging if someone passes `--tag` without a value locally. Not worth the added complexity for this POC.

File: tests/integration/fixtures/seed_etcd.sh [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    "namespace": "openshift-oauth-apiserver",
+    "labels": {"revision": "1"},
+    "annotations": {}
+  },
+  "spec": {
+    "template": {
+      "metadata": {"annotations": {}}
+    }
+  }
+}'
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply tests/integration/fixtures/seed_etcd.sh]: Intentional -- recert hardcodes apiversion "v1" in K8sResourceLocation::new for these deployments (see ocp_postprocess.rs lines 82-103). The etcd key path is derived from the apiversion, so using "apps/v1" would produce a different key and recert would not find the resource.

[the remark this answers is not in the document]
Comment (sebrandon1) [reply tests/integration/fixtures/seed_etcd.sh]: Already addressed — recert hardcodes apiversion "v1" in K8sResourceLocation::new for these deployments (ocp_postprocess.rs lines 82-103). The etcd key path derives from the apiversion; using "apps/v1" produces a different key path and recert would not find the resource.

**Comment by sebrandon1** [comment]
Related work for [CNF-26037](https://redhat.atlassian.net/browse/CNF-26037):
- [openshift-kni/lifecycle-agent#7725](https://github.com/openshift-kni/lifecycle-agent/pull/7725) — Unit tests for recert config generation, certificate utilities, and controller conditions

**Comment by sebrandon1** [comment]
@coderabbitai full review

File: Dockerfile.integration [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+WORKDIR /app
+RUN apt-get update && apt-get install -y protobuf-compiler
+COPY Cargo.toml Cargo.lock build.rs ./
+COPY src/ src/
+COPY vendor/ vendor/
+COPY .cargo/ .cargo/
+RUN cargo build --release --bin recert
+
+FROM docker.io/library/debian:bookworm
+WORKDIR /app
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply Dockerfile.integration:11]: This is a test-only container that runs in ephemeral CI environments and is never deployed. Adding a non-root user would add complexity (etcd permissions, tmp paths, fixture generation) with no security benefit for a disposable container. Deferring for now.

File: Dockerfile.integration [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+ARG ETCD_VERSION=v3.5.17
+
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    openssl openssh-client python3-yaml curl ca-certificates \
+    && curl -sL "https://github.com/etcd-io/etcd/releases/download/${ETCD_VERSION}/etcd-${ETCD_VERSION}-linux-amd64.tar.gz" | tar xz \
+    && mv "etcd-${ETCD_VERSION}-linux-amd64/etcd" /usr/local/bin/ \
+    && mv "etcd-${ETCD_VERSION}-linux-amd64/etcdctl" /usr/local/bin/ \
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply Dockerfile.integration]: Valid suggestion in general, but this is a test-only image downloading from GitHub releases over HTTPS in ephemeral CI. The checksum would also need updating on every etcd version bump. Low risk for this context - deferring.

File: tests/integration/scenarios/test_crypto_force_expire.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+EOF
+
+RECERT_CONFIG="${workdir}/config.yaml" run_recert_expect_success > /dev/null
+
+not_after_after=$(openssl x509 -noout -enddate -in "${crypto_dir}/ca.crt" 2>/dev/null | cut -d= -f2)
+not_after_epoch=$(date -d "$not_after_after" +%s)
+now_epoch=$(date +%s)
+
+# force_expire sets notAfter to approximately "now", allow 60s margin
+assert_lt "$not_after_epoch" "$((now_epoch + 60))" "force-expired cert should have notAfter near or in the past"
```
[the remark this answers is not in the document]
Comment (sebrandon1) [reply tests/integration/scenarios/test_crypto_force_expire.sh:22]: The 60s margin is intentional to avoid flaky tests in CI where timing can vary due to container startup and resource contention. Tightening to 5s would risk intermittent failures. The test already validates that force_expire produces a certificate with notAfter near the current time rather than years in the future.

**Comment by sebrandon1** [comment]
@omertuc Can this get an 'ok-to-test'?

**Comment by omertuc** [comment]
/override pull-ci-rh-ecosystem-edge-recert-main-e2e-aws-ovn-single-node-recert-parallel

**Comment by omertuc** [comment]
/override pull-ci-rh-ecosystem-edge-recert-main-e2e-aws-ovn-single-node-recert-serial

**Comment by sebrandon1** [comment]
Thanks @omertuc!
